### PR TITLE
Add CallbackHandle to pallet-assets

### LINKED ISSUE #12279: Callbacks for pallet-assets
### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Description of bug

Hi there,

it would be very useful if we could register callback methods for when new assets in `pallet-assets` are created, destroyed or manipulated in some other way.
The motivation behind is making assets cross-chain compatible for EVM contracts - in order to do this automatically, we need to do some actions after a new asset has been created. Being able to do this just using what `pallet-assets` provides would be great.

E.g. introduce a trait like:
```Rust
/// Callbacks for various asset manipulation functions
pub trait AssetsCallback<AssetId, AccountId> {

	/// Indicates that asset with `id` was successfully created by the `owner`
	fn created(id: &AssetId, owner: &AccountId);

	/// Indicates that asset with `id` has just been destroyed
    fn destroyed(id: &AssetId);
}
```

and have it as part of `pallet-assets` `Config`:
```Rust
	pub trait Config<I: 'static = ()>: frame_system::Config {
                ....
		/// Callback methods for asset state change (e.g. asset created or destroyed)
		type Callback: AssetsCallback<Self::AssetId, Self::AccountId>;
	}
```
Then it's just a matter of calling it once `create` or `destroy` have been successfully executed.
This would be a lightweight change, no benchmarks would really need to be changed if user doesn't introduce any custom hooks.

Is this a feasible change for `pallet-assets`?
Or should it just be handled by another pallet in chains that require it?

Related to [this](https://github.com/paritytech/substrate/discussions/12235) discussion basically.

### Steps to reproduce

_No response_

### PR DESCRIPTION
**Pull request by Dinonard**
This PR adds an option to register callbacks for `pallet-assets` which are invoked when a new asset is created or an existing one destroyed.

It's a simple change but requires a small runtime code update.
Benchmarks shouldn't be impacted in case default empty implementation is used.

cumulus companion: https://github.com/paritytech/cumulus/pull/1947

Closes #12279

**TODO**
- [x] unit test
- [x] Companion PR

### CODE REVIEW

**Comment by tonyalaribe** [comment]
@Dinonard Thanks for making this PR! 
I've thinking a bit about this and it raises two questions:

1. Is this not a usecase that could be covered by subscribing to the generated events?
2. How will these callbacks interact with extrinsic weights? For eg, what if the caller of the function does very expensive calculations or storage operations in the callback? Won't we be unable to generate a correct weight in that state? And won't that be a possible attack vector?

cc @joepetrowski @jsidorenko

**Comment by Dinonard** [comment]
Hi @tonyalaribe!
 
 > 1. Is this not a usecase that could be covered by subscribing to the generated events?
Event subscription is for interacting with the client, it's not part of the state transition function, so I don't believe this is applicable here.

> 2. How will these callbacks interact with extrinsic weights?
In case current `pallet-assets` users don't use this callback, they can just keep the weight as they are. In case they decide to introduce new custom logic, they have to rerun the benchmarks themselves and use the new generated weights instead.

File: frame/transaction-payment/asset-tx-payment/src/tests.rs
Comment (ggwpez) [inline frame/transaction-payment/asset-tx-payment/src/tests.rs:169]: A test would be nice.

**Review by ggwpez — COMMENTED** [review]
Seems fine to have.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
The job name - cargo-check-benches
The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1966720

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
The job name - cargo-check-benches
The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1966721

File: frame/transaction-payment/asset-tx-payment/src/tests.rs
Comment (Dinonard) [reply frame/transaction-payment/asset-tx-payment/src/tests.rs:169]: Added.

File: frame/assets/src/tests.rs
Comment (jsidorenko) [inline frame/assets/src/tests.rs:1006]: I would put `use sp_io::storage;` on the top and simply call `storage::get()`

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
The job name - cargo-check-benches
The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1969864

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
The job name - cargo-check-benches
The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1969865

File: frame/assets/src/tests.rs
Comment (Dinonard) [reply frame/assets/src/tests.rs]: Sure, changed.

**Review by jsidorenko — APPROVED** [review]
👍

File: frame/assets/src/functions.rs
Comment (tonyalaribe) [reply frame/assets/src/functions.rs:708]: I would also like to hear if this is possible. 
But I would assume it's not. I've only seen this done in tests.

File: frame/assets/src/tests.rs
Comment (tonyalaribe) [inline frame/assets/src/tests.rs:1012]: Maybe check that is actually true, and not just that it exists?

**Review by tonyalaribe — APPROVED** [review]

**Review by ggwpez — APPROVED** [review]

**Comment by ggwpez** [comment]
I put it into the audit queue.

File: frame/assets/src/tests.rs
Comment (Dinonard) [reply frame/assets/src/tests.rs:1012]: We could do that, the nature of the check is that it's `None` vs `Some`.
In that regard, perhaps just changing the type to `()` would be better.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2119553

**Comment by ggwpez** [comment]
> @ggwpez sorry if this is a silly question, but could you tell me if and what I can do next?
> 
> Should I prepare `cumulus` companion? I see that contributing guide references `polkadot` companions, doesn't mention `cumulus` though.

Sorry did not see.  
We are waiting for audit. I will give them another ping. Better not change any code anymore then.  
Your Cumulus companion looks good! Exactly how its done :laughing:

File: frame/assets/src/functions.rs
Comment (ggwpez) [reply frame/assets/src/functions.rs:758]: Yea probably a bad idea since it would need to read _all_ prior events.  
So it does currently not matter.

**Comment by ggwpez** [comment]
bot merge

**Comment by Dinonard** [comment]
Can this be merged? :slightly_smiling_face:

**Comment by ggwpez** [comment]
bot merge

**Comment by ggwpez** [comment]
> Can this be merged? slightly_smiling_face 

@Dinonard Well, i tried. You see the error message from the bot:  

> Error: Github API says "Allow edits from maintainers" is not enabled for https://github.com/paritytech/cumulus/pull/1947. The bot would use that permission to push the lockfile update after merging this PR. Please check https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork.

**Comment by Dinonard** [comment]
> @Dinonard Well, i tried. You see the error message from the bot:

Thanks!
Not sure what to make of that error but in `cumulus` companion it says:
> Rule "Runtime files" needs in total 2 DISTINCT approvals, but 1 were given.

Quite strange given that that there are two distinct approvals.

**Comment by joepetrowski** [comment]
> Quite strange given that that there are two distinct approvals.

You need another approval from someone in the `polkadot-review` group.

**Comment by ggwpez** [comment]
bot merge

**Comment by ggwpez** [comment]
You still did not update the Cumulus PR to be edit-able by maintainers @Dinonard

**Comment by hoonsubin** [comment]
> You still did not update the Cumulus PR to be edit-able by maintainers @Dinonard

Do you know how we can enable "allow edits by maintainers" from a fork that lives under an org account? [This thread](https://github.com/orgs/community/discussions/5634) suggests that that is not possible. Worst case, we'll have to close this PR and open a new one from an individual account fork

**Comment by ggwpez** [comment]
> …or you manually merge this.

Yea. Then we know this for the future.

**Comment by Dinonard** [comment]
> > …or you manually merge this.
> 
> Yea. Then we know this for the future.

Thanks!

I initially checked with @hoonsubin since I thought it might be due to me not having admin privileges under `AstarNetwork/cumulus`.

Should I modify the companion PR by updating the `.lock` file to point to latest master?

File: bin/node/runtime/src/lib.rs
Comment (gilescope) [inline bin/node/runtime/src/lib.rs:1481]: (Sorry am late to the party.), was just wondering if `OnCreated = (); OnDestroyed = ()` might make it a bit more similar to other pallet callbacks?

Comment (Dinonard) [reply bin/node/runtime/src/lib.rs:1481]: My idea was to keep it simple by only having one type but I do like your suggestion. It's more verbose :)

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v0-9-37/1736/1

**Pull request merged**
